### ISSUE TITLE
Gamma distribution does not have a CDF implementation, which is very useful in Pyro

### ISSUE THREAD

**Issue by bshabashFD** [pytorch/pytorch#41637]
## 🚀 Feature
Implement a CDF method for the Gamma distribution (torch.distributions.gamma.Gamma)

## Motivation

I'm trying to write Pyro code which uses censored data on a Gamma distribution, but I need the CDF of the Gamma distribution for that. The CDF has closed form solution so it should be easy to implement. Currently I need to cheat by using Scipy's gamma distribution and its CDF so that incurs run time 

## Pitch

I envision a new method added to `torch.distributions.gamma.Gamma` with the signature

```
def cdf(self, value):
# closed form solution code here
```

However, this solution would require the gamma function which I'm not sure is implemented (is torch._standard_gamma the gamma function?). If it isn't implemented I'd have to use `torch.exp(torch.lgamma(...))`

## Alternatives

Currently I'm using Scipy's gamma distribution CDF. It's not great since the 
(a) the signature is different, and
(b) detaching, converting to numpy, and then back to pytorch tensors takes up run time

cc @vincentqb @fritzo @neerajprad @alicanb @vishwakftw

**Comment by bshabashFD** [comment]
@gchanan , I see you added some editing to my issue, what is the process for getting the "ok, go for it?"

**Comment by fritzo** [comment]
@bshabashFD While Pytorch already implements a differentiable `torch.lgamma` you'll also need to implement a differentiable lower incomplete gamma function. IIRC the last effort to do so was #3877.

**Comment by bshabashFD** [comment]
@fritzo 
I'm trying to go through the PyTorch code and the topic you linked but I have no idea where to even get started looking at how lgamma is implemented to consider how to import the incomplete gamma function. Can you give me some pointers on which files to look at to understand the implementation hierarchy?

**Comment by cfinlay** [comment]
Hi all

Some info that may be helpful. I recently coded up a rudimentary upper incomplete gamma function G(s,x) in PyTorch, for some particular values of s. I followed [Boost's implementation](https://www.boost.org/doc/libs/1_61_0/libs/math/doc/html/math_toolkit/sf_gamma/igamma.html). I will say, that doing implementing this correctly, for all values of s, is a pretty big undertaking.

Moreover, I dug into differentiation of the incomplete gamma function, and it is seriously non-trivial to implement. The [derivative with respect to s](https://en.wikipedia.org/wiki/Incomplete_gamma_function#Derivatives) involves a rather nasty special function called the [Meijer G-function](https://en.wikipedia.org/wiki/Meijer_G-function). I was not able to find any standard numerical methods for computing the Meijer G-function.

So, if you do undertake this commendable task, it may be simpler if you only implement differentiation wrt x, which has a nice closed form expression.

**Comment by fritzo** [comment]
> @bshabashFD: Can you give me some pointers on which files to look at ... ?

I usually just grep around, for example you can find `digamma` (the derivative of `lgamma`) via:
```
ag --cpp -l digamma
```
then open the last match, `aten/src/ATen/native/Math.h`. I'm sure FB folks have more efficient habits.

**Comment by mfkasim1** [comment]
Hi, is there any updates on this? I'd like to take on this one if no one is working on it

**Comment by fritzo** [comment]
@mfkasim1 go ahead, to my knowledge no one is working on this.

**Comment by g-kakareko** [comment]
This is the implementation of the Gamma CDF. The derivative calculation is based on the series expansion described in: http://www.jstor.org/stable/2348014
```
import torch
from torch.autograd import gradcheck

def d_igamma_dp_series_expansion(p: torch.Tensor, x: torch.Tensor, eps: float = 1e-5, n_max: int = 100) -> torch.Tensor:
    """
    Algorithm AS 187: Derivatives of the Incomplete Gamma Integral
    Author(s): R. J. Moore
    Stable URL: http://www.jstor.org/stable/2348014

    Calculate d igamma(p,x) / dp using a series expansion (cf. Moore, 1982)
    This is valid for p <= x <= 1 and also for x < p.
    Notation from Moore 1982
    Args:
        p: actual observations
        x: model prediction
        eps: the minimal accepted accuracy when the series expansion is
        n_max: max number of iteration
    Returns:
        Calculate d igamma(p,x) valid for p <= x <= 1
    """
    log_x = torch.log(x)
    p_plus_1 = p + 1.0
    log_f = p * log_x - torch.lgamma(p_plus_1) - x
    f = log_f.exp()
    df_dp = f * (log_x - torch.polygamma(0, p_plus_1))

    C_old = torch.ones_like(p)
    d_C_old_dp = torch.zeros_like(p)
    S = C_old
    dS_dp = d_C_old_dp

    idx_notconv = torch.arange(0, p.numel(), dtype=torch.long)

    converged = False

    for n in range(1, n_max):
        ppn = torch.reciprocal(p + n)
        C_new = x * ppn * C_old
        d_C_new_dp = C_new * (torch.reciprocal(C_old) * d_C_old_dp - ppn)
        # update indices
        idx_notconv = torch.where(torch.abs(C_new) > S * eps)[0]

        if idx_notconv.numel() > 0:
            S[idx_notconv] = (S + C_new)[idx_notconv]
            dS_dp[idx_notconv] = (dS_dp + d_C_new_dp)[idx_notconv]
        else:
            converged = True
            break

        C_old = C_new
        d_C_old_dp = d_C_new_dp

    return S * df_dp + f * dS_dp


def d_igamma_dp_cf_expansion(p: torch.Tensor, x: torch.Tensor, eps: float = 1e-5, n_max: int = 100) -> torch.Tensor:
    """
    Algorithm AS 187: Derivatives of the Incomplete Gamma Integral
    Author(s): R. J. Moore
    Stable URL: http://www.jstor.org/stable/2348014

    Calculate d igamma(p,x) / dp using a continued fraction expansion (cf. Moore, 1982)
    This is valid for the entire input domain outside of { {p <= x <= 1} U {x < p} }.
    Notation from Moore 1982
    Args:
        p: actual observations
        x: model prediction
        eps: the minimal accepted accuracy when the series expansion is
        n_max: max number of iteration
    Returns:
        Calculate d igamma(p,x) valid for p <= x <= 1
    """
    log_x = torch.log(x)
    log_f = p * log_x - torch.lgamma(p) - x
    f = log_f.exp()
    df_dp = f * (log_x - torch.polygamma(0, p))

    A0 = torch.ones_like(p)
    B0 = x
    A1 = x + 1.0
    B1 = x * (2.0 - p + x)

    # init derivatives
    dA0_dp = torch.zeros_like(p)
    dA1_dp = torch.zeros_like(p)
    dB0_dp = torch.zeros_like(p)
    dB1_dp = -x

    S = torch.zeros_like(p)
    S_old = torch.zeros_like(p)
    S_temp = torch.zeros_like(p)
    dS_dp = torch.zeros_like(p)

    # tensor indices that haven't converged yet (initially all of them)
    idx_notconv = torch.arange(0, p.numel(), dtype=torch.long)

    converged = False

    for n in range(2, n_max):
        a, b = (n - 1) * (p - n), 2.0 * n - p + x
        A2 = b * A1 + a * A0
        B2 = b * B1 + a * B0
        dA2_dp = b * dA1_dp - A1 + a * dA0_dp + (n - 1) * A0
        dB2_dp = b * dB1_dp - B1 + a * dB0_dp + (n - 1) * B0

        S_old[idx_notconv] = (A1 / B1)[idx_notconv]
        S_temp[idx_notconv] = (A2 / B2)[idx_notconv]

        # update indices
        idx_notconv = torch.where(torch.abs(S_temp - S_old) > S_temp * eps)[0]

        if idx_notconv.numel() > 0:
            S[idx_notconv] = S_temp[idx_notconv]
            # TODO: refactor in the future - this might be numerically unstable
            dS_dp[idx_notconv] = (dA2_dp / B2 - S_temp * (dB2_dp / B2))[idx_notconv]
        else:
            converged = True
            break

        # update intermediates and partials
        A0, B0 = A1, B1
        A1, B1 = A2, B2
        dA0_dp, dB0_dp = dA1_dp, dB1_dp
        dA1_dp, dB1_dp = dA2_dp, dB2_dp

    return -S * df_dp - f * dS_dp


def d_igamma_dp(p: torch.Tensor, x: torch.Tensor, eps: float = 1e-5, n_max: int = 200) -> torch.Tensor:
    """
    Algorithm AS 187: Derivatives of the Incomplete Gamma Integral
    Author(s): R. J. Moore
    Stable URL: http://www.jstor.org/stable/2348014

    Calculate the derivative of the incomplete gamma function as the series expansion
    Notation from Moore 1982
    Args:
        p: actual observations
        x: model prediction
        eps: the minimal accepted accuracy when the series expansion is
        n_max: max number of iteration
    Returns:
        derivative of the incomplete gamma function
    """
    # uses the notation in (Moore, 1982)
    d_igamma = torch.zeros_like(p)
    idx = torch.logical_or(torch.logical_and(p <= x, x <= 1), x <= p)
    idx_series = idx.nonzero(as_tuple=True)[0]
    idx_continued_fraction = (~idx).nonzero(as_tuple=True)[0]
    if idx_series.numel() > 0:
        p_s, x_s = p[idx_series], x[idx_series]
        d_igamma[idx_series] = d_igamma_dp_series_expansion(p_s, x_s, eps, n_max)
    if idx_continued_fraction.numel() > 0:
        p_cf, x_cf = p[idx_continued_fraction], x[idx_continued_fraction]
        d_igamma[idx_continued_fraction] = d_igamma_dp_cf_expansion(p_cf, x_cf, eps, n_max)
    return d_igamma


class CustomIGamma(torch.autograd.Function):
    """
    PyTorch IGamma function implementation
    https://pytorch.org/docs/stable/distributions.html
    """

    def forward(self, a: torch.Tensor, z: torch.Tensor):  # type: ignore # pylint: disable=W0221
        cdf_ = torch.igamma(a, z)
        self.save_for_backward(a, z)
        return cdf_

    def backward(self, grad_output):  # pylint: disable=W0221
        a, z = self.saved_tensors
        d_igamma_a = d_igamma_dp(a, z)
        # The exact formula can be found:
        # https://www.wolframalpha.com/input/?i=d+GammaRegularized%5Ba%2C+0%2C+z%5D+%2F+dz
        d_igamma_z = (-z + (a - 1.0) * torch.log(z) - torch.lgamma(a)).exp()
        return (grad_output * d_igamma_a, grad_output * d_igamma_z)


class GammaCDF(torch.autograd.Function):
    """
    PyTorch Gamma distribution CDF implementation. This implementation solves the following issue raised on the PyTorch forum:
    https://github.com/pytorch/pytorch/issues/41637
    """

    def forward(self, x: torch.Tensor, k: torch.Tensor, psi: torch.Tensor):  # type: ignore # pylint: disable=W0221
        cdf_ = torch.igamma(k, x / psi)
        self.save_for_backward(k, x, psi)
        return cdf_

    def backward(self, grad_output, eps=1e-5):  # pylint: disable=W0221
        k, x, psi = self.saved_tensors
        a = k

        # Numerical stability
        x[x < eps] = eps
        z = x / psi

        d_igamma_a = d_igamma_dp(a, z, eps)
        d_igamma_z = (-z + a * torch.log(z) - torch.lgamma(a)).exp()
        return (grad_output * d_igamma_z / x, grad_output * d_igamma_a, -grad_output * d_igamma_z / psi)
```
Checking the Gradients 
```
n = 200
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
print("Device type: ", device)

##################################
# Testing the GammaCDF function  #
##################################
x = torch.rand(n, 1 , dtype=torch.double, requires_grad=True).to(device)*100
k = torch.rand(n, 1 , dtype=torch.double, requires_grad=True).to(device)*100
psi = torch.rand(n, 1 , dtype=torch.double, requires_grad=True).to(device)*100
test = gradcheck(GammaCDF.apply, (x, k, psi), eps=1e-6, atol=1e-4)
print("GammaCDF", test)
```

**Comment by cianeastwood** [comment]
Any plans to implement an inverse CDF function for the Gamma distribution (like in TensorFlow, [here](https://github.com/tensorflow/probability/blob/88d217dfe8be49050362eb14ba3076c0dc0f1ba6/tensorflow_probability/python/math/special.py#L766) )?

**Comment by fritzo** [comment]
@cianeastwood I believe `Gamma.icdf()` is blocked by implementation of `torch.special` function `gammaincinv()`, see tracking issue #50345

**Comment by cianeastwood** [comment]
@fritzo  I see, thanks for the update. Is there an open request/issue for Gamma.icdf()? Worth opening a new one and pointing to the blocking issue [#50345](https://github.com/pytorch/pytorch/issues/50345)? So that it is picked up when unblocked?

**Comment by fritzo** [comment]
@cianeastwood feel free to open a new issue on `Gamma.icdf()`

**Comment by krikru** [comment]
Is there currently a simple workaroud for this issue?

**Comment by krikru** [comment]
For anyone who is interested, the CDF of the a gamma distribution with a certain `concentration` and a certain `rate` can be calculated as

CDF(x) = `1 - gammaincc(concentration, rate * x)`

where gammaincc is defined in [torch.special](https://pytorch.org/docs/stable/special.html#torch.special.gammaincc). I haven't tried differentiating through this function though, so I don't know if/how that works.

**Comment by evanberkowitz** [comment]
> CDF(x) = `1 - gammaincc(concentration, rate * x)`

I tried using both this and gammainc to train the parameters of the gamma distribution but neither's parameters are differentiable (I didn't check x).  But given that the distribution's cdf method is non-existent this issue should be reopened.

**Comment by janeyx99** [comment]
cc @neerajprad @fritzo who are the right people for which to bring this under their radar?

**Comment by mfkasim1** [comment]
@evanberkowitz Hi. I'm the one who implemented the `gammainc` and `gammaincc` (it was under the name `igamma` and `igammac` before it was moved to `torch.special`).
The common notation for incomplete gamma function is: `gammainc(s, x)` (similarly for `gammaincc`).
The derivative for `x` is implemented because the expression is simple enough, but the derivative for `s` is not implemented.
The problem with the derivative with respect to `s` is that it involves [Meijer G-function](https://en.wikipedia.org/wiki/Meijer_G-function).
Implementing Meijer G-function will take a huge effort as it is one general function and I can't find any open source implementation that is reliable.
If you (or anyone) would like to implement the Meijer G-function in pytorch, it will be great!

**Comment by mfkasim1** [comment]
I also checked in pytorch 2.0.0, `torch.distributions.Gamma` does indeed have `cdf`:
```python
import torch.distributions as td
m = td.Gamma(torch.tensor([1.0]), torch.tensor([1.0]))
m.cdf(torch.tensor([0.5]))
```

**Comment by evanberkowitz** [comment]
Is the cdf differentiable wrt s in 2.0.0?  Or still not (because of Meijer G)?

**Comment by mfkasim1** [comment]
Still not, because of the Meijer G reason. I checked one time on JAX (if I remember it correctly) and they can compute the gradient although the speed is quite slow. They probably implement it with finite difference.

**Comment by tillahoffmann** [comment]
Should we close this because the gamma distribution has a CDF as of #89955 and the outstanding issue is that `gammainc` doesn't have gradients (as discussed in #80025)?

**Comment by janeyx99** [comment]
Yes, tracking the missing differentiability in #80025 and closing this one sounds good. Thank you @mfkasim1 and @tillahoffmann for the clarification.
